source-firestore/pull.go

```diff
@@ -115,27 +122,91 @@ type captureState struct {
 
 type resourceState struct {
 	ReadTime     time.Time
+	Backfill     *backfillState
 	bindingIndex uint32
 }
 
+type backfillState struct {
+	Cursor string    // The last document backfilled
+	MTime  time.Time // The UpdateTime of that document when we backfilled it
+}
+
+func (s *backfillState) Equal(x *backfillState) bool {
+	if s == nil {
+		return x == nil
+	} else if x == nil {
+		return false
```

Comment:
In order to deal properly with various edge cases around adding/removing bindings, capture state is tracked per resource path, but the Firestore APIs for streaming and querying data all operate in terms of collection groups based on the final path element. So during startup there's a bit of logic (see line 333) which amounts to:
1. Resume streaming from the oldest resume timestamp associated with a particular collection group (and then timestamp-based filtering on a per-resource basis will prevent duplicate already-processed events).
2. Also perform at most one backfill from a given collection group, resuming from the persisted state if it's the same for all resources in that group and resetting back to the start if there's a mismatch.

The `Equal()` helper method helps to make that logic slightly clearer, since `nil` is a valid backfill state meaning "no backfill is ongoing or necessary right now".

Reply:
Okay, I understand that it has value, but the naming is confusing. I would strongly expect `x.Equal(nil)` and `nil.Equal(x)` to be the same. I see that this is only used in one place, do you think we could rename this to something more suitable for its use?

Reply:
ah I see, you are right this is commutative, I got confused by the if conditions 👍🏽